### ISSUE TITLE
istio-cni: failed to remove pod from mesh

### ISSUE THREAD

**Issue by deshifcodes** [istio/istio#53829]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

**Namespaces previously added to ambient mesh are not deleted**

After removing the label (**istio.io/dataplane-mode**) from the **tools** namespace:

`kubectl label ns tools istio.io/dataplane-mode-`

the **tools** namespace continues to be in the ambient mesh:

`./istioctl ztunnel-config workload`

![2024-11-08 09 56 26](https://github.com/user-attachments/assets/02ccf617-bfd0-409e-aaf1-a7043a040ff7)

istio-cni logs:
![2024-11-08 10 09 58](https://github.com/user-attachments/assets/ae159d17-42d3-4b3a-9a4c-c0e8c7531303)

### Version

```Text
control plane version: 1.24.0
k8s: v1.30.4
CNI Plugin: cilium v1.16.3
```

**Comment by bleggett** [comment]
`failed to setup execution environment / failed to remount / permission denied` is likely the actual problem yea.

**Comment by deshifcodes** [comment]
Bare metal
OS Ubuntu server 24.04 server
Linux worker-01 6.8.0-48-generic #48-Ubuntu SMP PREEMPT_DYNAMIC Fri Sep 27 14:04:52 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

Another example. Adding a new namespace to the ambient mesh:

`kubectl label ns test istio.io/dataplane-mode=ambient`

istio-cni logs:
![2024-11-09 14 52 35](https://github.com/user-attachments/assets/4ebf9db0-5874-44cb-ae66-b8f882ee2b64)

`./istioctl ztunnel-config workload`

Even so, the namespace is added to the ambient mesh:

![2024-11-09 15 01 42](https://github.com/user-attachments/assets/31546a3a-540c-4d3e-bc17-e582c03e962c)

**Comment by keithmattix** [comment]
Are you running with seLinux configured by any chance?

**Comment by deshifcodes** [comment]
Ubuntu server was installed as standard. Selinux is not installed

**Comment by deshifcodes** [comment]
The following information may be useful. When I add a namespace to ambient mesh, the following logs appear in journalctl:

```
...
Nov 10 12:41:29 worker-03 kernel: audit: type=1400 audit(1731206489.637:2139): apparmor="DENIED" operation="ptrace" class="ptrace" profile="cri-containerd.apparmor.d" pid=2167 comm="install-cni" requested_mask="read" denied_mask="read" peer="unconfined"
Nov 10 12:41:29 worker-03 kernel: audit: type=1400 audit(1731206489.637:2140): apparmor="DENIED" operation="ptrace" class="ptrace" profile="cri-containerd.apparmor.d" pid=2167 comm="install-cni" requested_mask="read" denied_mask="read" peer="unconfined"
Nov 10 12:41:29 worker-03 kernel: audit: type=1400 audit(1731206489.637:2141): apparmor="DENIED" operation="ptrace" class="ptrace" profile="cri-containerd.apparmor.d" pid=2167 comm="install-cni" requested_mask="read" denied_mask="read" peer="unconfined"
...
```

I did not configure AppArmor separately. This configuration is the default in Ubuntu Server 24.04

**Comment by keithmattix** [comment]
This is super helpful, thanks for sharing! Yeah it looks like AppArmor doesn't want istio-cni to have access to that directory despite it having CAP_SYS_PTRACE capabilities

**Comment by howardjohn** [comment]
What I'm not certain of is exactly what is failing and why. Some failures may trigger other failures. I know the mount operation is blocked by the appArmor I think, but not sure that is required

**Comment by keithmattix** [comment]
From my quick reading (so definitely not authoritative), reading from procfs is what's causing app Armor to log there in journalctl

**Comment by deshifcodes** [comment]
Edited Daemonset by adding the Unconfined profile for AppArmor:

![2024-11-08 09 56 26](https://github.com/user-attachments/assets/8bdb2a82-a3be-4313-ad0f-00bc0a9dffd8)

Next, I add the **test** namespace to the ambient mesh (there are no errors and journalctl is empty)

istio-cni logs:

![2024-11-11 14 31 57](https://github.com/user-attachments/assets/591bab08-f81a-43d4-bd2b-032721089780)

Now if I delete the test namespace from the ambient mesh, I get errors and the namespace is not deleted from the mesh:

![2024-11-11 14 36 43](https://github.com/user-attachments/assets/58769e01-2395-48a7-a531-763e1db60f46)

**Comment by howardjohn** [comment]
> From my quick reading (so definitely not authoritative), reading from procfs is what's causing app Armor to log there in journalctl

https://github.com/containerd/containerd/blob/2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41/contrib/apparmor/template.go#L71

Shouldn't this only block write to procfs? Not read? It would not surprise me if there is something weird

`deny mount` is suspicious though.

**Comment by keithmattix** [comment]
Yeah according to that config, it should only block write. The journalctl logs point to a blocked read which is strange to me

**Comment by bleggett** [comment]
> Edited Daemonset by adding the Unconfined profile for AppArmor:
> 
> ![2024-11-08 09 56 26](https://private-user-images.githubusercontent.com/155072573/384786496-8bdb2a82-a3be-4313-ad0f-00bc0a9dffd8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzNDk3MzksIm5iZiI6MTczMTM0OTQzOSwicGF0aCI6Ii8xNTUwNzI1NzMvMzg0Nzg2NDk2LThiZGIyYTgyLWEzYmUtNDMxMy1hZDBmLTAwYmMwYTlkZmZkOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExMVQxODIzNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYzU2NDY0NDc0YmE1ODQzOWQ1YzM1MTViZTdjZTY5MWY1ZmEzNmU2M2Y2NTNiNTdiMTA1ZDc1MTRmY2I1MzM5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.oFpCfVJDWzvgzlnNKDgtIYmv4GOfZ6tqwuO893MF7Lo)
> 

QQ - does setting `privileged: true` change any behavior here? It may elide/bypass AppArmor.

I find it odd, if this is a default Ubuntu, that no one would have hit this in 1.23.

**Comment by howardjohn** [comment]
On slack someone said privileged=true fixed it

**Comment by bleggett** [comment]
> On slack someone said privileged=true fixed it

I was afraid of that. Okay.

**Comment by keithmattix** [comment]
Shipping our own profile would but rough too because Kubernetes enforces that the profile exists before applying the pod. So we'd have to work around the race there

**Comment by howardjohn** [comment]
Do we know why setting to unconstrained doesn't work?

**Comment by keithmattix** [comment]
Yeah folks said unconstained doesn't work. I wonder if the issue is this ptrace role in the appArmor config: https://github.com/containerd/containerd/blob/2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41/contrib/apparmor/template.go#L94. I have zero context here, but maybe allowing ptracing containers within the pod _disallows_ ptracing containers/processes outside the pod (similar to how Istio AuthZ policies work)

**Comment by bleggett** [comment]
> Yeah folks said unconstained doesn't work. I wonder if the issue is this ptrace role in the appArmor config: https://github.com/containerd/containerd/blob/2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41/contrib/apparmor/template.go#L94. I have zero context here, but maybe allowing ptracing containers within the pod _disallows_ ptracing containers/processes outside the pod (similar to how Istio AuthZ policies work)

Yeah https://gitlab.com/apparmor/apparmor/-/wikis/TechnicalDoc_Proc_and_ptrace#apparmor-3-with-ptrace-rules implies that a bit.

**Comment by jaellio** [comment]
Would setting the `appArmorProfile` to unconfined update the ptrace role in the appArmor config to be `ptrace (trace, read) peer=unconfined`? Or does the `profile` remain `cri-containerd.apparmor.d`?

**Comment by howardjohn** [comment]
Maybe a dumb question, but do we actually know what is blocked? Or just
"read" of something? I never saw anything indicating *what*

**Comment by keithmattix** [comment]
I think multiple things are failing looking back at the original screenshot. The istio-cni error is "failed to remount /: permission denied" which points to that mount deny flag in the default (unconstrained) apparmor policy.

**Comment by jaellio** [comment]
> Maybe a dumb question, but do we actually know what is blocked? Or just "read" of something? I never saw anything indicating *what*

Working on replicating this to do a more detailed analysis

**Comment by keithmattix** [comment]
Ah, what if you remove the appArmor profile with privileged: true?

**Comment by howardjohn** [comment]
> Ah, what if you remove the appArmor profile with privileged: true?

in slack someone already confirmed this fixes it
If we cannot disable appArmor only, we should default to privileged and allow non-apparmor users to opt out

**Comment by bleggett** [comment]
> allow non-apparmor users to opt out

My only concern with this is the same one I had before - making it a user-facing knob is easy to misunderstand/abuse if we aren't careful - I want to avoid people getting in to the trap of thinking "I can make `istio-cni` unprivileged if I set `privileged=false`" - which they _can never_ do. It is _always_ a privileged pod, in reality and as per the Kubernetes PSS.

If we let people opt-out, I'm very much not clear how we would message that in a way that wouldn't confuse the hell out of 99% of the people looking at that flag.

Maybe exposing it as `values.appArmorBypass=true|false` would be enough. I get that this means we have to document a special flag for ~every Ubuntu tho, which is not ideal.

**Comment by bleggett** [comment]
> I agree that making privileged a user facing nob would reintroduce the confusion we were trying to avoid. A `values.appArmorBypass` flag makes sense to me. However, how many users would install istio not knowing apparmor is enabled and spend extra cycles debugging?

Yeah - I think we would have to default `appArmorBypass` to `true` (to echo what @howardjohn said above)

If you _know_ you aren't using appArmor, you can opt-in to setting it to false.

**Comment by jaellio** [comment]
> > Maybe a dumb question, but do we actually know what is blocked? Or just "read" of something? I never saw anything indicating _what_
> 
> Working on replicating this to do a more detailed analysis

I was finally able to replicate this yesterday and I think there are two things going on. Setting the appArmorProfile to unconfined on the istio-cni-node daemonset does resolve the ptrace failures. The pods still fail to be removed from the mesh because of failures related to deleting nonexistent raw iptable rules. 

- https://github.com/istio/istio/pull/53888 Should resolve the AppArmor failures.
- https://github.com/istio/istio/pull/53153 Should resolve the iptable failures.

**Comment by leosarra** [comment]
> * [Idempotency and reconciliation for cni iptables #53153](https://github.com/istio/istio/pull/53153) Should resolve the iptable failures.

My commit is a prereq to fix it as it adds the reconciliation feature but it is not enough on its own. @bleggett is working on a follow-up to that commit to trigger reconciliation on all existing pods.

**Comment by howardjohn** [comment]
If it's feasible to have a smaller fix for the patch only id love to do it.
Unfortunately I don't have any ideas though

**Comment by ppalucki** [comment]
I expirience this issue on completly different setup (kind version 0.25.0 with default CNI k8s 1.31.2)
 
Changing label namespace from `istio.io/dataplane-mode=ambient` to `istio.io/dataplane-mode=none` removes iptables rules but does not remove from HBONE

to reproduce:
```bash
# create namespace and add it to mesh
kubectl create ns nethost
kubectl label namespace auth istio.io/dataplane-mode=ambient --overwrite
# run iptables-save/ss -tpln in one terminal 
kubectl run --rm -it nethost --image=nicolaka/netshoot --restart=Never --privileged -n nethost -- watch -n1 'iptables-save; ss -tpln'
# run istioctl ztunnel-config workload |grep nethost in another terminal
watch -n1 'istioctl ztunnel-config workload |grep nethost'
```

then run alternately

```
kubectl label namespace auth istio.io/dataplane-mode=ambient --overwrite
```
and try to remove
```
kubectl label namespace auth istio.io/dataplane-mode=none --overwrite
```

adding usually works without issue, but removing iptables are cleared but proxy stays and ztunnel-config workloads that nethost is still in a HBONE
but after removing:

```
Every 1.0s: iptables-save; ss -tpln                                                                                                                                                                                                     

# Generated by iptables-save v1.8.10 (nf_tables) on Fri Nov 15 13:58:29 2024
*mangle
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
COMMIT
# Completed on Fri Nov 15 13:58:29 2024
# Generated by iptables-save v1.8.10 (nf_tables) on Fri Nov 15 13:58:29 2024
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
COMMIT
# Completed on Fri Nov 15 13:58:29 2024
State                             Recv-Q                            Send-Q                                                       Local Address:Port                                                         Peer Address:Port                            Process
LISTEN                            0                                 128                                                              127.0.0.1:15053                                                             0.0.0.0:*
LISTEN                            0                                 128                                                                  [::1]:15053                                                                [::]:*
LISTEN                            0                                 128                                                                      *:15008                                                                   *:*
LISTEN                            0                                 128                                                                      *:15006                                                                   *:*
LISTEN                            0                                 128                                                                      *:15001                                                                   *:*
```
and
```
Every 1.0s: istioctl ztunnel-config workload |grep nethost
nethost                   nethost                                               10.244.0.80 kind-control-plane None     HBONE
```

it looks like some race condition, because it sometimes happens and sometimes not

FYI:
```
# istioctl version
client version: 1.23.2
control plane version: 1.24.0
data plane version: 1.24.0 (1 proxies)
```

situation is even worse for two or more containers :(
first one being remove (iptables/proxy)
```
nethost                   nethost                                               10.244.0.80 kind-control-plane None     TCP
nethost                   nethost1                                              10.244.0.84 kind-control-plane None     HBONE
nethost                   nethost2                                              10.244.0.83 kind-control-plane None     HBONE
```
and two others have iptables cleared but proxy stays and reports HBONE

:(


should I report this as another ticket ?

**Comment by leosarra** [comment]
@ppalucki I suspect that the error you have in the logs about inpod rules not being deleted correctly should be addressed by my commit, that error led to an early exit in the RemovePodFromMesh() when certain iptables cfgs are used.

I noticed the same problem while working on the idempotency/reconciliation PR while writing the test coverage so I tried to fix it.
From the description of my PR:
> - Resolved in-pod deletion errors caused by the deletion logic incorrectly assuming that certain commands would work, even when the relevant rules were not present. The creation of these rules depends on the values of tproxy and ingressMode. Therefore, certain failed deletion commands cannot reliably indicate successful deletion. See https://github.com/istio/istio/commit/1b03f56e1476cf850afe11f70abdb15ed810e686.

On the other end, I cannot say for sure why sometimes you experienced the issue and sometimes you did not as I'm not well versed in the cni-agent logic.

**Comment by howardjohn** [comment]
This issue has diverged a bit but as far as I can tell everything discussed here has been solved now. Thanks everyone!

If you see similar issues please open a new issue and make sure you are on the latest versions.
